### ISSUE TITLE
GPU version gives nan value

### ISSUE THREAD

**Issue by adonisues** [HawkAaron/warp-transducer#17]
Hi. First thank you for your code.
I  have a problem GPU version warp-transducer.
I applied gpu version warp-transducer to your E2E-ASR code. (with minor adjust) 
At this case, loss became "nan" after several steps.
Is there any requirement in cuda or another problem?

thank you

pytorch 1.0.1
ubuntu 18
cuda 10

**Comment by HawkAaron** [comment]
@adonisues 
I used PyTorch 1.0 with CUDA 9.0
CUDA10 may not support.

**Comment by guangsenw** [comment]
I am using the latest pytorch 1.0 with CUDA 9.0, same problem with nan loss after several steps. I think there may be some bugs inside the kernel code gpu_rnnt_kernel.h. If I replace this file with the one from the mxnet repo (https://github.com/HawkAaron/mxnet-transducer/blob/master/rnnt_include/detail/gpu_rnnt_kernel.h), the loss seems to be normal

**Comment by HawkAaron** [comment]
Please rebuild it with `cmake -DUSE_NAIVE_KERNEL=on ..`, there are truly some bugs inside `gpu_rnnt_kernel.h`, I will fix it asap.

**Comment by HawkAaron** [comment]
The bug is fixed.

**Comment by adonisues** [comment]
Thank you for your fix. It works well.
But there is a typo  at "/warprnnt_pytorch/__init__.py", line 79"
"labels = torch.cat([label[:i] for label, i in zip(labels, ylen)], dim=0)" : ylen ==> label_lens
And this line make labels 1D tensor if batch size is 1, so error occurs at certify_inputs.

By the way, I have a question about the included softmax.
 Is there any reason you use your own softmax for GPU version instead of nn.functional.log_softmax ?

**Comment by guangsenw** [comment]
> The bug is fixed.

Thanks for the fix. 
in the new warprnnt_pytorch/__init__.py, RNNTLoss seems to miss the backward function causing autograd to throw a "NotImplementedError". BTW, does the new commit also fix the bug in gpu_rnnt_kernel.h, i.e., the bug in the non-naive computation of alphas and betas? Thanks

**Comment by HawkAaron** [comment]
Thanks for your bug report, it works well now.

BTW, the GPU version has some tricks on GPU memory consuming, it only store two values for each prediction (blank and label).  I have moved log_softmax before loss function call.

**Comment by adonisues** [comment]
Hello GPU version is working now.
But CPU version is not working when batch size > 1.
It makes Segmentation fault.

**Comment by HawkAaron** [comment]
CPU version works well now.
